add draft documentation for On Demand PATCH request (#143)
livepeer/livepeer-studio-docs: docs/references/vod/patch.mdx

>**NOTE:** 
- Livepeer Studio's defaut fields for `metadata` are:
```json
    "name": , //Populated with the asset name
    "description": , //A default description referring to the asset name above
    "image": //A default image including the Livepeer logo as a thumbnail
    "animation_url": //URL for the Livepeer player on IPFS playing the video from Livepeer Studio
    "external_url": //URL for the Livepeer player hosted on lvpr.tv playing the video from Livepeer Studio
    "properties": {
      "video": //CID of the original video file saved on IPFS
      "com.livepeer.playbackId": //playback ID of the asset in Livepeer Studio
    }
```

>**NOTICE** that you can change the value or delete any of these default fields by providing a value in the `nftMetadata` field. You can delete any of the above fields by specifying them as `null` in your overrides.

Example: `"animation_url": null`

### Request

```bash
curl -X PATCH https://livepeer.studio/api/asset/{id} \
-H 'content-type: application/json' \
-H 'authorization: Bearer {api-key}' \
-d {  "meta": {
        "title": "My awesome video",
        "description": "This is a video of my awesome life",

>         "tags": ["awesome", "life", "video"],

Reviewer: This cannot be an array. Must have only string values.

Author: @victorges 
So -->  "tags": "awesome", "life", "video"

Reviewer: That wouldn't be a valid JSON tho. You'd need to have a single string and handle splitting on the application logic. e.g.
```suggestion
        "tags": "awesome,life,video",
```

On the app that uses this, they'd ned to do sth like `asset.meta.tags.split(",")` to get the tags list (don't need to document that tho, just sharing more context).

Author: See commit below ⬇️

Reviewer: Still invalid. Suggested a tool that you can use to write valid JSONs.

---
title: PATCH updating storage
---

`PATCH /asset/:id -d {"storage": {"ipfs": {}}` can be used to modify any `asset` that was stored with On Demand. It is a single endpoint that can update only the following mutable properties of an asset:

- `name`
- `storage`

> - `meta`

Reviewer: We should have a page [like this](https://docs.livepeer.studio/references/stream/overview) to explain each field of the asset objects (not a change for this PR, but a note for the future). We shouldn't have fields documented scattered among multiple docs, but have a single reference documentation with all of the asset fields (then we can also link to it in places where we would explain such fields).

Author: Agreed

Author: Made an issue for this https://github.com/livepeer/livepeer-studio-docs/issues/147

## Update Stored Asset

> > **NOTE:** When updating the asset information, the `storage` object must be included in the body of the request

Reviewer: We should updated the API to make this not necessary. Users should be able to update each property independently.

## Updating Asset Name

This will allow the name of the stored asset in Livepeer Studio to be changed.

### Request

```bash
curl -X PATCH https://livepeer.studio/api/asset/{id} \
-H 'content-type: application/json' \
-H 'authorization: Bearer {api-key}' \
-d {
  "name": "New name"

>   "storage": {"ipfs": {}}

Reviewer: When we fix the API, remove this
```suggestion
```

  }
```

> ## Updating Storage

Reviewer: We need instructions on how to track the progress of the storage update. Maybe we should have a separate guide for it tho, since it's less of an API reference and more of a "how to" guide.

This allows to add additional storage location of where the assets are to be stored.

>**NOTE:** 

> - Currently assets cannot be deleted from storage if stored on Livepeer Studio. 

Reviewer: ```suggestion
- Currently assets cannot be deleted from the default Livepeer Studio storage. You can customize any additional storage though.
```

> - Only IPFS is support as an additional storge option at the moment

Reviewer: ```suggestion
- Only IPFS is supported as additional storage at the moment. We are currently working on support for other decentralized storages like Arweave and Storj.
```

### Request

```bash
curl -X PATCH https://livepeer.studio/api/asset/{id} \
-H 'content-type: application/json' \
-H 'authorization: Bearer {api-key}' \

> -d {"storage": {"ipfs": {}}}

Reviewer: We should add some instructions about the `nftMetadata` field, which can be used to customize the asset NFT metadata.

Author: Need to know what are the metadata fields and how to customize it.

Reviewer: These are the default metadata templates: https://github.com/livepeer/task-runner/blob/4d1ae1768c01e979969bf7f00b81267db05e299c/task/export.go#L154
(default is the first one)

But there is no enforced schema here. Users can set anything they want. Each marketplace has their own standard, but they generally follow OpenSea's: https://docs.opensea.io/docs/metadata-standards
These are Lens, compatible and strongly based on OpenSea: https://docs.lens.xyz/docs/metadata-standards

All those marketplaces also make sure to be compatible with one of the real "NFT standards":
 - [ERC-721](https://eips.ethereum.org/EIPS/eip-721#specification) (basically `name`, `description` and `image`)
 - [ERC-1155](https://eips.ethereum.org/EIPS/eip-1155#metadata) (`name`, `description`, `decimals`, `image` and `properties`)

Also notice that the `attributes` field, used by OpenSea, Lens, etc, is not present in any stoken standard. It was just made up by OpenSea because ERC-721 didn't have it and kinda became the standard nowadays... We use the `properties` field by default, but most apps actually use `attributes` (including Lens).

We do not have any NFT metadata standards of our own. It is acceptable to say that users should just customize their NFT metadata following any standard that they want.

```

## Updating Metadata

This allows for modifying the metadata of the asset.

### Request

```bash
curl -X PATCH https://livepeer.studio/api/asset/{id} \
-H 'content-type: application/json' \
-H 'authorization: Bearer {api-key}' \
-d {

>   "storage": {"ipfs": {}},

Reviewer: ```suggestion
```

### Request

```bash
curl -X PATCH https://livepeer.studio/api/asset/{id} \
-H 'content-type: application/json' \
-H 'authorization: Bearer {api-key}' \
-d { "name": "New name" }
```

## Updating Storage

This allows to add additional storage location of where the assets are to be stored.

>**NOTE:** 
- Currently assets cannot be deleted from the default Livepeer Studio storage. You can  customize any additional storage though. 
- Only IPFS is supported as an additional storge at the moment. We are currently working on support for other decentralized storages like Arweave and Storj.

### Request

```bash
curl -X PATCH https://livepeer.studio/api/asset/{id} \
-H 'content-type: application/json' \
-H 'authorization: Bearer {api-key}' \
-d {"storage": {"ipfs": {}}}
```

## Updating Metadata

This allows for modifying the metadata of the asset, which is customizable NFT. Since each marketplace has their own standard for the `metatdata` of the NFT, it is up to you to determine which standard to use and fields to include.

>**NOTE:** 
- Livepeer Studio's defaut fields for `metadata` are:
```json
    "name": ,
    "description": ,

>     "cid": 

Reviewer: Ops, had sent a comment on the commit by mistake. Sending it again on the PR here.

This is not the right field name, it is actually `animation_url` and `properties.video`, following ERC-1155.

You should also add some instructions on how to override/delete these default fields, as the curr CLi print as well:
https://github.com/livepeer/video-nft/blob/main/src/cli/args.ts#L123-L126

Can also use the SDK docs for some further reference: https://livepeer.github.io/video-nft/classes/minter.Api.html#exportToIPFS

And finally, can resort to the implementation here to see the (default) metadata in detail, which can be overridden by what is set here: https://github.com/livepeer/task-runner/blob/4d1ae1768c01e979969bf7f00b81267db05e299c/task/export.go#L155

Let me know if you still need any additional info. I know that these are not very structured pieces of documentation so I can help clarify things further.

Reviewer: Also, upon revisiting the last link I sent with the documentation, I noticed we're populating a few other fields by default. The extensive list with some short description (iow, my take on this doc):

> - `name`: Populated with the asset name
> - `description`: A default description referring to the asset name above
> - `image`: A default image including the Livepeer logo as a thumbnail
> - `animation_url`: URL for the Livepeer player on IPFS playing the video from Livepeer Studio
> - `external_url`: URL for the Livepeer player hosted on lvpr.tv playing the video from Livepeer Studio
> - `properties["video"`: CID of the original video file saved on IPFS
> - `properties["com.livepeer.playbackId"]`: playback ID of the asset in Livepeer Studio
>
> Notice that you can change the value or delete any of these default fields by providing a value in the `nftMetadata` field. You can delete any of the above fields by specifying them as `null` in your overrides.

-d {  "meta": {
        "title": "My awesome video",
        "description": "This is a video of my awesome life",

>         "tags": "awesome","life","video"

Reviewer: This is still not valid JSON. You can any online tool to validate these. If you throw that on https://jsoneditoronline.org/ you'll get the errors.

As I've mentioned, this has to be a single string, not multiple strings separated by commas. That's not a valid JSON unless this field was an array.
